src/dune_rules/pkg_rules.ml

```diff
@@ -514,6 +516,65 @@ module Substitute = struct
 end
 
 module Run_with_path = struct
+  module Output : sig
+    type error
+
+    val io : error -> Process.Io.output Process.Io.t
+
+    val with_error
+      :  accepted_exit_codes:int Predicate.t
+      -> pkg:Dune_pkg.Package_name.t * Loc.t
+      -> display:Display.t
+      -> (error -> 'a)
+      -> 'a
+
+    val prerr : rc:int -> error -> unit
+  end = struct
+    type error =
+      { pkg : Dune_pkg.Package_name.t * Loc.t
+      ; filename : Dpath.t
+      ; io : Process.Io.output Process.Io.t
+      ; accepted_exit_codes : int Predicate.t
+      ; display : Display.t
+      }
+
+    let io t = t.io
+
+    let with_error ~accepted_exit_codes ~pkg ~display f =
+      let filename = Temp.create File ~prefix:"dune-pkg" ~suffix:"stderr" in
+      let io = Process.Io.(file filename Out) in
+      let t = { filename; io; accepted_exit_codes; display; pkg } in
+      let result = f t in
+      Temp.destroy File filename;
+      result
+    ;;
+
+    let error_msg ~is_critical t error =
```

Comment:
the `is_critical` bool is a bit weird, I wonder if that would be clearer to pass the message instead

Reply:
This issue is that in one case it's a `Warning` and in another one it's an `Error`. We need to stick to the behaviour in `Process` and raise only on critical error. I can change it by adding a `kind` instead of  `is_critical` using variant.

Reply:
Ah ok I missed that part. My main issue was the `critical` bit, indeed. What you're suggesting is fine by me, but as a final suggestion, maybe we can use `User_warning.emit` and the `is_error` flag, with just `Logs for package X` as message in both cases. I wonder if that would be enough?

Reply:
As it would display `Error` and `Warning` in front of the message, I think his is enough for now :+1: